### ISSUE TITLE
BlockAdapter namespace resolution is wrong when StorageNamespace is a prefix under the bucket

### ISSUE THREAD

**Issue by itaiad200** [treeverse/lakeFS#2397]
**Example**
Namespace: `gs://foo/bla`, Key: `bar/baz`

QualifiedKey:
Expected: StorageNamespace: `foo`, Key: `bla/bar/baz`
Actual: StorageNamespace: `foo`, Key: `/bla/bar/baz`

In some block storages like gs, this will result in the following object: `gs://foo//bla/bar/baz`. In S3, it doesn't make a difference and the object path is `gs://foo/bla/bar/baz`.

Since the adapter is consistent in the way it is referencing the object, we did not see an impact on usability when communicating with the lakeFS server solely.
BUT when trying to use other clients like the metadata client, which uses the physical address returned by the API (`gs://foo/bla/bar/baz`), operations fail with `NoSuchObject`.

Any fix to this issue MUST consider backward compatibility of existing repos which already stored according to the double-slash pattern.

**Comment by shabtai** [comment]
"Any fix to this issue MUST consider backward compatibility of existing repos which already stored according to the double-slash pattern."

do you want me to write tests that will verify the "backward compatibility" contract?

**Comment by itaiad200** [comment]
Thanks @shabtai for the kind offer!
What exactly do you have in mind?

**Comment by shabtai** [comment]
You probably need to test all your operations against storages that contain
objects created in ALL prior versions(?)

**Comment by itaiad200** [comment]
@shabtai We don't really know yet how to resolve it, I'm working on a design now and will let you know how you'll be able to assist once completed.

**Comment by itaiad200** [comment]
Did some mapping investigation for the blockstorage type, prefix existence and trailing slash existence. The mapping is between the storage namespace and the actual physical address:

#### No prefix, no trailing slash (gcp, azure, s3):
gs://bucket <> bucket/dummy
https://itailakefs.blob.core.windows.net/bucket <> bucket//dummy -> **problem**
s3://bucket <> bucket/dummy

#### No prefix, with trailing slash:
gs://bucket/ <> bucket/dummy
https://itailakefs.blob.core.windows.net/bucket/ <> bucket//dummy -> **problem**
s3://bucket/ <> bucket/dummy

#### With prefix, no trailing slash:
gs://bucket/prefix <>bucket//prefix/dummy -> **problem**
https://itailakefs.blob.core.windows.net/bucket/echad <> bucket/echad/dummy
s3://bucket/prefix <> bucket/prefix/dummy

#### With prefix, with trailing slash:
gs://bucket/prefix/ <> bucket//prefix/dummy -> **problem**
https://itailakefs.blob.core.windows.net/bucket/prefix/ <> bucket/prefix//dummy -> **problem**
s3://bucket/prefix/ <> bucket/prefix/dummy

3 Alternatives for solving this issue:

1. Change nothing in lakeFS server which will continue to work like it does today. Clients (like Metadata-client and lakeFSFS) should apply special logic to handle the 12 different options above **FOREVER**. 
Pros: No data migration needed, no special code in the lakeFS side needed.
Cons: Maintaining the compatibility with the bug in the client side.

2. Fix the bug in lakeFS and advance the commit/range version number. Take into account that version number when computing the physical address in both the server and the clients **FOREVER**. 
Pros: No data migration needed.
Cons: Maintaining the compatibility with both versions in the server and client side. 

3. Fix the bug in lakeFS and perform a **full data migration**. Some repositories will need to move all of their data, some will need to move none. lakeFS is probably not the right choice for these sorts of ETL jobs, so we'll need to provide a Spark script for the migration.
Pros: No need to be backward compatible here. Once a repo is migrated, it would work with server and clients without any special handling. Migration isn't needed for any S3 repo, which is the vast majority of users.
Cons: An ETL job of migrating the entire data handled by a lakeFS repo is painful. Failing half-way through is something that is going to be really hard to recover from. Serious downtime for big repos. No way to migrate down.

**Comment by asm0dey** [comment]
Potentially with option 3 spark is not the only and maybe not the best option. Each store has its respective clients with the support of `move`. And the info on what should be moved may be obtained from the lakeFS server if I understand correctly.

**Comment by itaiad200** [comment]
@asm0dey Spark is indeed not the only option for these sorts of tasks. But isn't this what Spark does under the hood? Performing `Rename` on an object calls the appropriate client move operation (Need to validate that's the case for gs/Azure FS). Since there's the potential of moving millions of objects, I'm not that interested in writing the orchestration/distribution logic myself, that's why Spark came into place. 
Not sure if a communication with the lakeFS server is needed, given the storage namespace, the job can infer the input and output base paths.

**Comment by asm0dey** [comment]
Absolutely, it's what Spark does under the hood. And now it suddenly stroke me that there is pyspark with which it is easy to write such an application.

**Comment by arielshaqed** [comment]
# Another alternative

We can run the fix in phases.  I believe this has operational advantages, and does not significantly increase the amount or risk of the coding risk.  I split it into 4 phases, but we might merge some of them if we don't care about the advantages after any one of them

## Phase 1: Fix the future

* When looking up a file, try using the correct ("new") logic.  If that lookup fails, try using the existing incorrect ("old") logic.
  - In particular, _do not add a version number_ to metadata files.  So IIUC we never need to rewrite those. :-)
* When writing a file, write it to the correct location.  (This might be controlled by a flag, to allow users to choose _when_ the upgrade starts.  E.g. test _other_ features of the new version before switching the flag and committing to having no easy downgrade path.)
* There is no "new" version of Graveler or staging data, just a _different_ ("correct") interpretation of what is already there.
* (Add good monitoring for this, to allow users to determine how many double lookups there are.)

After phase 1, the damage is limited to the past.

## Phase 2: Dynamic migration

_This phase is optional, one could skip it and move to the next phase._

* Whenever a lookup to the correct location fails, attempt to migrate the file to its new location.
  - (Graveler files move but their contents need not change!)
  - We can use true "move" operations here, which should be fast(-er)
  - We _must_ have atomic per-object operations, so works for 
* Can apply any operations budget here, e.g. limit to X operations per minute: if we _don't_ migrate, nothing bad happens, we just continue to have this slow lookup.
* Can delete incorrect file locations as part of the same operation ("move") or after successfully migrating everything ("copy").

With phase 2, double lookups on frequently accessed objects decrease, so the double-lookup code rarely runs.  Its runtime cost decreases, but the code still exists.

## Phase 3: Backfill

* Use some task orchestrator to migrate _all_ the files.
* Ops advantages on (1):
   - Requires no downtime because (1) is still in place.
   - Requires no single successful run because (1) is still in place.
* Can delete incorrect file locations as part of the same operation ("move") or after successfully migrating everything ("copy").

This removes all incorrect ("old") file locations.

## Phase 4: Remove dead code

* Delete the old incorrect (second-chance) code

This reduces the maintenance burden.

## Advantages

This phased approach:

1. Gives (distinct) value after each phase
2. Allows us to stop without doing all of the phases (i.e., we can wait until after phase
3. Requires no downtime
